### PR TITLE
Netwayste Reliability (Part 3)

### PR DESCRIPTION
**Pull request by Manghi**
This is quite a large PR as I branched off in October, so take as much time as you need for a proper review. The other two parts were merged within #5 and #6 .

**Some pertinent highlights**
* `(net)` Removed the concept of a separate transmission and receive queue traits and merged them into a single trait `NetQueue`.
* `(net)` Added a macro to handle most udp packet sending as `netwayste_send!()`.
* `(net|client|server)` Added explicit handlers to joining/leaving rooms and logging in (instead of just using `OK`) because of the out-of-order nature of this problem. 
  * (Another side-effect was removing `ClientState::last_request_action`.)
* `(net)` Retransmission of un-acknowledged packets, with retry counters and an aggressive mode
* `(client|server)` Created additional tokio streams for network-specific events
* `(net|client|server)` Finally added proper logging via `env_logger`. 

Feel free to use the `testnet.sh` shell script. It was written/tested on my system but it should be generic enough to work on yours... You'll ~~likely~~ hopefully only need to update the terminal it launches. I would reduce `roomCmdCount=500` to something reasonable like 100 first.

The overall performance isn't where I want it to be at, but I think we should tackle that problem in a few weeks after integration and some other more pressing features are attended to.

---

You'll need to run with the exports listed below otherwise the logging includes mio-specific trace statements. `env_logger` has a way to specify the logging level, so I was able to turn it off for `futures`, and `tokio_core`. It did not work for `mio` so I still need to find out why. Might have something to do with `mio` not being imported explicitly? The `RUST_LOG` env var works in the mean time. 
```Rust
    env_logger::Builder::new()
    .(/*omitted*/)
    .filter(None, LevelFilter::Trace)   // Worked
    .filter(Some("futures"), LevelFilter::Off)   // Worked
    .filter(Some("tokio_core"), LevelFilter::Off)   // Worked
    .filter(Some("mio"), LevelFilter::Off) // did not work so I never kept it
    .init();
``` 
```Bash
export RUST_LOG=server # For Server
export RUST_LOG=client # For client
cargo run --bin [server|client]
```

### CODE REVIEW

File: notes/linux_tc.txt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# Classes need to have the same major number as their parent.
+#   The following command instantly creates classes 1:1, 1:2, and 1:3.
+
+#tc qdisc add dev eth0 root handle 1: prio
+
+
+# 1:3 refers to the third band of the PRIO queue.
+# This described by qdisc ID of 30:
+
+#tc class add dev lo parent 1:3 classid 1:10 netem delay 200ms 20ms distribution normal    # Delayed 200
```
Comment (AaronM04) [inline notes/linux_tc.txt:12]: Huh, interesting! Good research. Where do you get these `classid 1:XX` numbers from?

**Comment by AaronM04** [comment]
Where does this come from? It's a lot of output, and drowns out the important stuff.
```
Sun 2019-03-10 18:37:17.610142 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:18.370336 [TRACE] - event Readable | Writable Token(0)
Sun 2019-03-10 18:37:18.370442 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:18.595280 [TRACE] - event Writable Token(0)
Sun 2019-03-10 18:37:18.595355 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:19.326751 [TRACE] - event Readable | Writable Token(0)
Sun 2019-03-10 18:37:19.326821 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:19.589742 [TRACE] - event Writable Token(0)
Sun 2019-03-10 18:37:19.589783 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:20.448348 [TRACE] - event Readable | Writable Token(0)
Sun 2019-03-10 18:37:20.448428 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:20.590752 [TRACE] - event Writable Token(0)
Sun 2019-03-10 18:37:20.590827 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:21.329304 [TRACE] - event Readable | Writable Token(0)
Sun 2019-03-10 18:37:21.329394 [TRACE] - loop process - 1 events, 0.000s
Sun 2019-03-10 18:37:21.591626 [TRACE] - event Writable Token(0)
Sun 2019-03-10 18:37:21.591687 [TRACE] - loop process - 1 events, 0.000s
```

**Comment by Manghi** [comment]
> Where does this come from? It's a lot of output, and drowns out the important stuff.
> 
> ```
> ... omitted ...
> ```

Read the second half of the initial PR comment, under the horizontal line. It covers this. Briefly, it comes from `mio` but I couldn't find a way to turn it off without scoping the RUST_LOG env var to the specific binary (client/server).

File: src/server.rs [hunk 2]
```diff
@@ -30,13 +30,15 @@ extern crate semver;
 #[cfg(test)]
 #[macro_use]
 extern crate proptest;
+extern crate chrono;
 
+#[macro_use]
 mod net;
 
-use net::{RequestAction, ResponseCode, Packet, LineCodec, UniUpdateType, BroadcastChatMessage};
+use net::{RequestAction, ResponseCode, Packet, LineCodec, UniUpdateType, BroadcastChatMessage, NetworkManager, NetworkQueue};
```
Comment (AaronM04) [inline src/server.rs:38]: This line is a bit long for my tastes.

File: src/net.rs [hunk 3]
```diff
@@ -39,13 +38,46 @@ pub const VERSION: &'static str = env!("CARGO_PKG_VERSION");
...
 // Instead of returning that index, I return this much larger value and avoid insertion into the queues.
 // (110 is the avg weight of an amino acid in daltons :] Much larger than our current queue size)
 const MATCH_FOUND_SENTINEL: usize = 110;
 
+//////////////// Public Macros /////////////////
+
+#[macro_export]
+macro_rules! netwayste_send {
+    // Client
+    ($_self:ident, $tx:expr, $dest:expr, ($failmsg:expr $(, $args:expr)*)) => {
```
Comment (AaronM04) [inline src/net.rs:58]: `$_self:ident` is unused. That's the only difference between Client and Server sections. Why not just have one for both?

File: src/client.rs [hunk 4]
```diff
@@ -84,36 +87,35 @@ impl ClientState {
...
     fn in_game(&self) -> bool {
         self.room.is_some()
     }
 
     fn disconnecting(&self) -> bool {
-        if let Some(ref action) = self.last_req_action {
-            return action == &RequestAction::Disconnect
-        }
-        false
+        self.disconnect_initiated
```
Comment (AaronM04) [inline src/client.rs:118]: Might be better to just check `self.disconnect_initiated` directly rather than having this function. It's a more descriptive name anyway :)

File: src/client.rs [hunk 5]
```diff
@@ -129,40 +131,83 @@ impl ClientState {
...
+                    // a Response packet, the target identifier is the `request_ack`.
+
+                    // Only process responses we haven't seen
+                    if self.response_sequence <= sequence {
+                        trace!("RX Buffering: Resp.Seq.: {}, {:?}", self.response_sequence, packet);
+                        // println!("TX packets: {:?}", self.network.tx_packets);
+                        // None means the packet was not found so we've probably already removed it.
+                        if let Some(_) = self.network.tx_packets.remove(&packet)
+                        {
+                            let _ = self.network.rx_packets.buffer_item(packet);
```
Comment (AaronM04) [inline src/client.rs:207]: I don't think we actually need the `let _ =` here.

File: src/client.rs [hunk 6]
```diff
@@ -458,15 +489,27 @@ fn print_help() {
 
 //////////////////// Main /////////////////////
 fn main() {
-    drop(env_logger::init());
+    env_logger::Builder::new()
+    .format(|buf, record| {
+        writeln!(buf,
+            "{} [{:5}] - {}",
```
Comment (AaronM04) [inline src/client.rs:495]: Looks nice! I like this log line formatting thing.

File: src/client.rs [hunk 7]
```diff
@@ -458,15 +489,27 @@ fn print_help() {
...
+            Local::now().format("%a %Y-%m-%d %H:%M:%S%.6f"),
+            record.level(),
+            record.args(),
+        )
+    })
+    .filter(None, LevelFilter::Trace)
+    .filter(Some("futures"), LevelFilter::Off)
+    .filter(Some("tokio_core"), LevelFilter::Off)
```
Comment (AaronM04) [inline src/client.rs:503]: Maybe adding another line for `mio` is a way to deal with #17?

File: src/client.rs [hunk 8]
```diff
@@ -526,9 +570,18 @@ fn main() {
             Event::UserInputEvent(user_input)
         }).map_err(|_| ());
 
+    let network_stream = stream::iter_ok::<_, io::Error>(iter::repeat( () ));
+    let network_stream = network_stream.and_then(|_| {
```
Comment (AaronM04) [inline src/client.rs:574]: Why not `tick_stream` and `Event::TickEvent`?

File: src/net.rs [hunk 9]
```diff
@@ -438,78 +510,74 @@ pub trait NetworkQueue<T: Ord+Sequenced> {
...
-type ItemQueue<T> = VecDeque<T>;
-
-pub struct TXQueue {
-    pub queue: ItemQueue<Packet>,
-}
-
-pub struct RXQueue<T> {
-    pub queue: ItemQueue<T>,
-    pub buffer_wrap_index: Option<usize>
+pub struct Timestamp {
```
Comment (AaronM04) [inline src/net.rs:524]: Can we think of another name for this? A timestamp is just a point in time. This is something more.

File: src/net.rs [hunk 10]
```diff
@@ -746,15 +874,97 @@ impl NetworkManager {
...
+         udp_tx: &mpsc::UnboundedSender<(SocketAddr, Packet)>,
+         addr: SocketAddr,
+         confirmed_ack: Option<u64>,
+         indices: &Vec<usize>) {
+
+        let mut error_occurred = false;
+        let mut failed_index = 0;
+
+        // Retransmit all packets after that are still in the queue after RETRANSMISSION_THRESHOLD_IN_MS
+        for index in indices.iter() {
```
Comment (AaronM04) [inline src/net.rs:898]: It's a little clearer to me to use `for &index in indices.iter()` and then just refer to it in the block as `index` instead of `*index`.

File: src/server.rs [hunk 11]
```diff
@@ -82,7 +88,7 @@ struct Player {
     cookie:        String,
     addr:          SocketAddr,
     name:          String,
-    request_ack:   Option<u64>,          // most recent request sequence number received
+    request_ack:   Option<u64>,          // most recent request sequence number received. This +1 is what we'd next expect.
```
Comment (AaronM04) [inline src/server.rs:91]: Maybe reword: "The next number we expect is `request_ack + 1`" ? :)

File: src/server.rs [hunk 12]
```diff
@@ -329,6 +335,16 @@ impl ServerState {
         opt_player.unwrap()
     }
 
+    fn get_player_mut(&mut self, player_id: PlayerID) -> &mut Player {
```
Comment (AaronM04) [inline src/server.rs:338]: Hell yeah! :tada:

File: src/server.rs [hunk 13]
```diff
@@ -563,34 +578,261 @@ impl ServerState {
...
+        if self.rooms.len() == 0 {
+            return;
+        }
+
+        for room in self.rooms.values() {
+            if room.player_ids.len() == 0 {
+                continue;
+            }
+
+            for player_id in &room.player_ids {
```
Comment (AaronM04) [inline src/server.rs:769]: Same suggestion as in other file: prefer `for &player_id in ...` and then reference with `player_id` instead of `*player_id`.

File: src/server.rs [hunk 14]
```diff
@@ -563,34 +578,261 @@ impl ServerState {
...
+        } else {
+            // request_ack has not been set yet, likely first packet
+            player.request_ack = Some(0);
+            true
+        }
+    }
+
+    /// Processes a player's request action for all non Connect requests. If necessary, a response is buffered
+    /// for later transmission
+    fn process_player_request_action(&mut self, player_id: PlayerID, action: RequestAction) -> Result<Option<Packet>, Box<Error>> {
```
Comment (AaronM04) [inline src/server.rs:625]: Maybe it's just me, but I got a little confused by the long chain of function calls, each handling only one type of message.
process_player_request_action -> handle_request_action -> process_request_action

**Review by AaronM04 — APPROVED** [review]
Phenomenal! I'm going to go ahead and merge this. My comments can be addressed in follow-up PRs if you want.

Also, it would be nice to have info in the readme on how to change the log level using environment variables.

**Comment by AaronM04** [comment]
Thanks @Manghi ! :boom: :+1:

**Comment by AaronM04** [comment]
@Manghi one last thing: can you make a PR for the parent repo (including the `Cargo.lock` changes)?

File: notes/linux_tc.txt (hunk 1 above)
Comment (Manghi) [reply notes/linux_tc.txt:12]: It's assigning an ID for the class that's set up when the qdisc is configured like on line 6.

File: src/net.rs (hunk 3 above)
Comment (Manghi) [reply src/net.rs:58]: I think I initially had that for the client-side which had the network stuff in its parent trait `ClientState`. Sure.

File: src/server.rs (hunk 14 above)
Comment (Manghi) [reply src/server.rs:625]: Definitely not just you. Will look into clarifying that call-chain.

File: src/net.rs (hunk 9 above)
Comment (Manghi) [reply src/net.rs:524]: What about `NetAttempt`?

File: src/client.rs (hunk 7 above)
Comment (Manghi) [reply src/client.rs:503]: I tried that but it appeared to have no effect; probably didn't do it correctly so I'll try one more thing. I like your suggestion of asking on `freenode#mio`, haven't done that yet but will do soon.

File: src/client.rs (hunk 8 above)
Comment (Manghi) [reply src/client.rs:574]: I wanted to be able to send the keep-alives at a different rate than all of the other TX network traffic. Currently they're both set to run every second but I don't think those are the right values.
